test/performance/scale_from_zero_test.go

```diff
@@ -0,0 +1,242 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package performance
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"testing"
+	"time"
+
+	pkgTest "github.com/knative/pkg/test"
+	"github.com/knative/pkg/test/logging"
+	"github.com/knative/test-infra/tools/testgrid"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/knative/serving/test"
+)
+
+const (
+	configName                       = "perftest-config"
+	routeName                        = "perftest-route"
+	ScaleFromZeroAvgTestGridProperty = "perf_ScaleFromZero_Average"
+	ScaleFromZeroTestGridClass       = "ScaleFromZero"
+)
+
+const (
+	helloWorldExpectedOutput = "Hello World!"
+)
+
+type testContext struct {
+	logger         *logging.BaseLogger
+	deploymentName string
+	names          test.ResourceNames
+	domain         string
+}
+
+type stats struct {
+	avg time.Duration
+}
+
+func setupClients(t *testing.T) *test.Clients {
+	clients, err := test.NewClients(
+		pkgTest.Flags.Kubeconfig,
+		pkgTest.Flags.Cluster,
+		test.ServingNamespace)
+	if err != nil {
+		t.Fatalf("Couldn't initialize clients: %v", err)
+	}
+	return clients
+}
+
+func tearDown(clients *test.Clients, names test.ResourceNames, logger *logging.BaseLogger) {
+	if clients != nil && clients.ServingClient != nil {
+		clients.ServingClient.Delete([]string{names.Route}, []string{names.Config}, []string{names.Service})
+	}
+}
+
+func createRouteAndConfig(clients *test.Clients, logger *logging.BaseLogger, imagePath string, options *test.Options) (test.ResourceNames, error) {
+	var names test.ResourceNames
+	names.Config = test.AppendRandomString(configName, logger)
+	names.Route = test.AppendRandomString(routeName, logger)
+
+	if _, err := test.CreateConfiguration(logger, clients, names, imagePath, options); err != nil {
+		return test.ResourceNames{}, err
+	}
+	_, err := test.CreateRoute(logger, clients, names)
+	return names, err
+}
+
+func isDeploymentScaledToZero() func(d *v1beta1.Deployment) (bool, error) {
+	return func(d *v1beta1.Deployment) (bool, error) {
+		return d.Status.ReadyReplicas == 0, nil
+	}
+}
+
+func setup(clients *test.Clients, name string) (*testContext, error) {
+	logger := logging.GetContextLogger("TestScaleFromZero (" + name + ")")
+	imagePath := test.ImagePath("helloworld")
+	names, err := createRouteAndConfig(clients, logger, imagePath, &test.Options{})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create Route and Configuration: %v", err)
+	}
+	test.CleanupOnInterrupt(func() { tearDown(clients, names, logger) }, logger)
+
+	logger.Info("Waiting for route to be marked ready.")
+	if err := test.WaitForRouteState(clients.ServingClient, names.Route, test.IsRouteReady, "RouteIsReady"); err != nil {
+		return nil, fmt.Errorf("The Route %s was not marked as Ready to serve traffic: %v", names.Route, err)
+	}
```

Comment:
Using `test.CreateRunLatestServiceReady` should give us this state already + it should make the Config.Get + Route.Get superfluous as it already returns these objects.